src/subscriptions.rs

```diff
@@ -1,40 +1,279 @@
 	message_tx: Sender<(Header, Instant)>,
 	error_sender: Sender<anyhow::Error>,
 ) {
-	async fn subscribe_and_process(
-		rpc_client: avail::Client,
-		message_tx: Sender<(Header, Instant)>,
-	) -> Result<()> {
-		let mut new_heads_sub = rpc_client.blocks().subscribe_finalized().await?;
-
-		while let Some(message) = new_heads_sub.next().await {
-			let received_at = Instant::now();
-			if let Ok(block) = message {
-				let header = block.header().clone();
-				info!(header.number, "Received finalized block header");
-				let message = (header, received_at);
-				if let Err(error) = message_tx.send(message).await.context("Send failed") {
-					error!("Fail to process finalized block header: {error}");
-				}
-			}
-		}
-		Err(anyhow!("Finalized blocks subscription disconnected"))
-	}
-
-	if let Err(error) = subscribe_and_process(rpc_client, message_tx).await {
+	if let Err(error) = subscribe_check_and_process(rpc_client, message_tx).await {
 		error!("{error}");
 		if let Err(error) = error_sender.send(error).await {
 			error!("Cannot send error to error channel: {error}");
 		}
 	}
 }
+
+#[derive(Debug, Encode)]
+enum SignerMessage {
+	_DummyMessage(u32),
+	PrecommitMessage(Precommit),
+}
+
+#[derive(Clone, Debug, Decode, Encode, Deserialize)]
+struct Precommit {
+	pub target_hash: H256,
+	/// The target block's number
+	pub target_number: u32,
+}
+
+#[derive(Clone, Debug, Decode, Deserialize)]
+struct SignedPrecommit {
+	pub precommit: Precommit,
+	/// The signature on the message.
+	pub signature: Signature,
+	/// The Id of the signer.
+	pub id: EdPublic,
+}
+#[derive(Clone, Debug, Decode, Deserialize)]
+struct Commit {
+	pub target_hash: H256,
+	/// The target block's number.
+	pub target_number: u32,
+	/// Precommits for target block or any block after it that justify this commit.
+	pub precommits: Vec<SignedPrecommit>,
+}
+
+#[derive(Clone, Debug, Decode)]
+struct GrandpaJustification {
+	pub round: u64,
+	pub commit: Commit,
+	pub _votes_ancestries: Vec<Header>,
+}
+
+impl<'de> Deserialize<'de> for GrandpaJustification {
+	fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+	where
+		D: serde::Deserializer<'de>,
+	{
+		let encoded = bytes::deserialize(deserializer)?;
+		Self::decode(&mut &encoded[..])
+			.map_err(|codec_err| D::Error::custom(format!("Invalid decoding: {:?}", codec_err)))
+	}
+}
+
+#[derive(Clone, Debug)]
+enum Messages {
+	Justification(GrandpaJustification),
+	ValidatorSetChange((Vec<EdPublic>, u64)),
+	NewHeader(Header, Instant),
+}
+async fn subscribe_check_and_process(
+	subxt_client: Client,
+	message_tx: Sender<(Header, Instant)>,
+) -> Result<()> {
+	let mut header_subscription = subxt_client
+		.rpc()
+		.subscribe_finalized_block_headers()
+		.await?;
+	// Get the hash of the head (finalized)
+	let last_finalized_block_hash = rpc::get_chain_head_hash(&subxt_client).await?;
+
+	// Current set of authorities, implicitly trusted, fetched from grandpa runtime.
+	let mut validator_set =
```

Comment:
Yes and no. We change nodes on startup in case there are multiple. But also, changes in the validator set are part of a header and that header is always signed off by previous validator set, so if we assume implicit trust in the first validator set (the one active on startup), we don't need to trust the node anymore.